evm-types.md

```diff
@@ -660,13 +652,22 @@ Addresses
     rule #addr(W) => W %Word pow160
 ```
 
--   `#lookup` looks up a key in a map and returns 0 if the key doesn't exist, otherwise returning its value.
+-   `#lookup*` looks up a key in a map and returns 0 if the key doesn't exist, otherwise returning its value.
+    It also makes sure the returned value is in the correct bitwidth, adjusting it if not.
 
 ```k
-    syntax Int ::= #lookup ( Map , Int ) [function]
- // -----------------------------------------------
-    rule [#lookup.some]: #lookup( (KEY |-> VAL:Int) M, KEY ) => VAL
-    rule [#lookup.none]: #lookup(                   M, KEY ) => 0 requires notBool KEY in_keys(M)
+    syntax Int ::= #lookup        ( Map , Int ) [function, functional, smtlib(lookup)]
+                 | #lookupMemory  ( Map , Int ) [function, functional, smtlib(lookupMemory)]
+ // ----------------------------------------------------------------------------------------
+    rule [#lookup.some]:       #lookup( (KEY |-> VAL:Int) M, KEY ) => VAL requires          0 <=Int VAL andBool VAL <Int pow256
+    rule [#lookup.outOfRange]: #lookup( (KEY |-> VAL:Int) M, KEY ) => 0   requires notBool (0 <=Int VAL andBool VAL <Int pow256)
+    rule [#lookup.none]:       #lookup(                   M, KEY ) => 0   requires notBool KEY in_keys(M)
+    rule [#lookup.notInt]:     #lookup( (KEY |-> VAL    ) M, KEY ) => 0   requires notBool isInt(VAL)
+
+    rule [#lookupMemory.some]:       #lookupMemory( (KEY |-> VAL:Int) M, KEY ) => VAL requires          0 <=Int VAL andBool VAL <Int pow256
+    rule [#lookupMemory.outOfRange]: #lookupMemory( (KEY |-> VAL:Int) M, KEY ) => 0   requires notBool (0 <=Int VAL andBool VAL <Int pow256)
+    rule [#lookupMemory.none]:       #lookupMemory(                   M, KEY ) => 0   requires notBool KEY in_keys(M)
+    rule [#lookupMemory.notInt]:     #lookupMemory( (KEY |-> VAL    ) M, KEY ) => 0   requires notBool isInt(VAL)
```

Comment:
Two questions:
1. What is the difference between #lookup and #lookupMemory here?
2. Where does the range condition come from?  (It wasn't there before right?)

Reply:
`#lookup` ensures that it will return something between `0` and `pow256`.

`#lookupMemory` ensures that it will return something between `0` and `256`.

Reply:
Oh, then the condition for lookupMemory needs to be fixed.

Reply:
Ah yeah, copy-paste error.